### PR TITLE
Support p384 curves

### PR DESCRIPTION
**Pull request by jacob-delgado**
**Please provide a description of this PR:**

Support P384 curves in addition to P256. Use P256 by default.

### CODE REVIEW

File: security/pkg/pki/util/crypto.go [hunk 1]
```diff
@@ -130,14 +131,14 @@ func GetRSAKeySize(privKey crypto.PrivateKey) (int, error) {
 	return pkey.N.BitLen(), nil
 }
 
-// IsSupportedECPrivateKey is a predicate returning true if the private key is EC based
-func IsSupportedECPrivateKey(privKey *crypto.PrivateKey) bool {
-	switch (*privKey).(type) {
+// GetElipticCurve is a predicate returning true if the private key is EC based
```
Comment (bleggett) [inline security/pkg/pki/util/crypto.go:134]:
```
// GetElipticCurve is a predicate returning true if the private key is EC based
```

nit - no longer correct

File: security/pkg/pki/util/crypto.go [hunk 2]
```diff
@@ -130,14 +131,14 @@ func GetRSAKeySize(privKey crypto.PrivateKey) (int, error) {
...
+func GetElipticCurve(privKey *crypto.PrivateKey) (elliptic.Curve, error) {
```
Comment (bleggett) [inline security/pkg/pki/util/crypto.go:135]:
```suggestion
func GetEllipticCurve(privKey *crypto.PrivateKey) (elliptic.Curve, error) {
```

nit: sp

File: security/pkg/pki/util/generate_cert.go [hunk 3]
```diff
@@ -42,9 +42,17 @@ import (
 // to be used in key generation (e.g. ECDSA or ED2551)
 type SupportedECSignatureAlgorithms string
 
+// SupportedElipticCurves are the types of curves
+// to be used in key generation (e.g. P256, P384)
+type SupportedElipticCurves string
```
Comment (bleggett) [inline security/pkg/pki/util/generate_cert.go:47]:
```suggestion
type SupportedEllipticCurves string
```
nit: sp

File: security/pkg/pki/util/generate_csr.go [hunk 4]
```diff
@@ -45,7 +45,16 @@ func GenCSR(options CertOptions) ([]byte, []byte, error) {
 	if options.ECSigAlg != "" {
 		switch options.ECSigAlg {
 		case EcdsaSigAlg:
-			priv, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+			var curve elliptic.Curve
+			switch options.ECCCurve {
+			case P384Curve:
+				curve = elliptic.P384()
+			case P256Curve:
+				fallthrough
```
Comment (bleggett) [inline security/pkg/pki/util/generate_csr.go:53]:
```suggestion
			case P256Curve:
				fallthrough
```

mega-nit: can we be consistent with default case + P256 default handling in all these case statements?

e.g. either 

```
case
case 256
  fallthrough
default
  256
```

OR

```
case
case 256
  256
default
  256
```

but not a mix-and-match?

**Review by bleggett — APPROVED** [review]

File: releasenotes/notes/ecc-p384.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+apiVersion: release-notes/v2
+kind: feature
+area: security
+issue: []
+releaseNotes:
+- |
+  **Added** support for using P384 curves when using ECDSA
```
Comment (costinm) [inline releasenotes/notes/ecc-p384.yaml:7]: Can you add few more words on how a user would use this or link to a doc ?

File: security/pkg/pki/ca/ca.go [hunk 6]
```diff
@@ -364,8 +365,15 @@ func (ca *IstioCA) GenKeyCert(hostnames []string, certTTL time.Duration, checkLi
 	// use the type of private key the CA uses to generate an intermediate CA of that type (e.g. CA cert using RSA will
 	// cause intermediate CAs using RSA to be generated)
 	_, signingKey, _, _ := ca.keyCertBundle.GetAll()
-	if util.IsSupportedECPrivateKey(signingKey) {
+	curve, err := util.GetEllipticCurve(signingKey)
+	if err == nil {
 		opts.ECSigAlg = util.EcdsaSigAlg
+		switch curve {
+		case elliptic.P384():
+			opts.ECCCurve = util.P384Curve
```
Comment (costinm) [inline security/pkg/pki/ca/ca.go:373]: Not sure I understand - any reason to restrict this to only 384 ?

**Review by costinm — APPROVED** [review]
Would be great to have a doc ( wiki at least ) with how this whole thing works from a user perspective. And also we need at some point to discuss what is the proper API surface and long term plans ( in context of 'safe mode' where we want to tighten things up). 

IMHO cert provisioning should be a platform and K8S-wide feature, or use things like
CertManager.  I don't mind if our own 'fallback' CA adds more features, but we should avoid leaking that into the API and UX.

**Comment by jacob-delgado** [comment]
Long term people should be using a tool like cert-manager. It works with KeyFactor, Google CAS, Let's Encrypt, etc... That is the route that we are pushing our customers towards as it provides something more robust. However, it would be a good idea for us to consider how our users can configure their CSRs. Not so much for spiffee, but for TTLs and type of encryption used. I hate having a set of environment variables for all of these things. I had worked on something before, but I wasn't as familiar as I am with the set of requirements users are starting to ask for.

**Pull request merged**